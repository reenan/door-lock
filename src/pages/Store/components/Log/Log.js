import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Segment, Header } from 'semantic-ui-react'

import './Log.scss'

class Log extends Component {

  getFormattedDate = (date) => {
    date = new Date(date)

    const day = date.getUTCDate()
    const month = date.getUTCMonth()
    const year = date.getUTCFullYear()
    const hour = date.getUTCHours()
    const minute = date.getUTCMinutes()

    return <span>{day}/{month}/{year} {hour}:{minute}</span>
  }

  getFormattedEmployee = (employeeID, roleID) => {
    const { employees, roles } = this.props

    let result = []

    if (employees[employeeID]) {
      result.push(<span key={0}>{employees[employeeID].name}</span>)
    } else {
      result.push(<span key={0} className='deleted'>Deleted Employee</span>)
    }

    if (roles[roleID]) {
      result.push(<span key={1}> ({roles[roleID].name})</span>)
    } else {
      result.push(<span key={1} className='deleted'> (Deleted Role)</span>)
    }

    return result
  }

  getFormattedDoor = (doorID) => {
    const { doors } = this.props

    if (doors[doorID]) {
      return <span>{doors[doorID].name}</span>
    } else {
      return <span className='deleted'>Deleted Door</span>
    }
  }

  getFormattedAcceptedRequest = (isAllowed) => {
    if (isAllowed) {
      return <span className='green'>accepted</span>
    } else {
      return <span className='red'>not accepted</span>
    }
  }

  getFormattedString = (requestID) => {
    const request = this.props.requests[requestID]

    const dateElement = this.getFormattedDate(request.date)
    const employeeElement = this.getFormattedEmployee(request.employee, request.role)
    const doorElement = this.getFormattedDoor(request.door)
    const allowedElement = this.getFormattedAcceptedRequest(request.allowed)

    return (
      <p>
        On {dateElement}, {employeeElement} requested to open {doorElement} and
        his/hers request was {allowedElement}
      </p>
    )
  }

  render() {
    const { requestsKeys } = this.props

    return (
      <div>
        <Header inverted size='small'>
          Here you can check all requests that were made to open your doors.
        </Header>
        <Segment tertiary className='logs-container'>
          {
            requestsKeys.length > 0 ?
              requestsKeys.map(requestID => (
                <div key={requestID}>
                  { this.getFormattedString(requestID) }
                </div>
              )) :
              <p>Looks like no request was made yet. Go and try one!</p>
          }
        </Segment>
      </div>
    )
  }
}

Log.propTypes = {
  requests: PropTypes.object.isRequired,
  doors: PropTypes.object.isRequired,
  roles: PropTypes.object.isRequired,
  employees: PropTypes.object.isRequired,
  requestsKeys: PropTypes.array.isRequired,
}

export default Log
