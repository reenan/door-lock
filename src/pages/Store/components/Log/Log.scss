.logs-container {
  max-height: 500px;
  overflow-y: scroll;

  > div {
    margin: 1em 0;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }

    p {
      span {

        &.deleted {
          color: gray;
          font-style: italic;
        }

        &:not(.deleted) {
          font-weight: bold;
        }

        &.red {
          color: #d65757;
        }

        &.green {
          color: #1c8145;
        }
      }
    }
  }
}
