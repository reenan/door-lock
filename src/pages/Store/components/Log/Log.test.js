import React from 'react'
import EnzymeToJson from 'enzyme-to-json'
import { mount } from 'enzyme'

import Log from './Log'

it('renders without crashing', () => {
  const component = mount(
    <Log
      requests={{}}
      doors={{}}
      roles={{}}
      employees={{}}
      requestsKeys={[]}
    />
  )

  expect(EnzymeToJson(component)).toMatchSnapshot()
})


it('renders without crashing with request and saved data', () => {
  const mockDate = new Date('11/12/1994')
  
  const component = mount(
    <Log
      requests={{
        mock: {
          date: mockDate,
          employee: 'mockEmployee',
          role: 'mockRole',
          door: 'mockDoor',
          allowed: true,
        }
      }}
      doors={{
        mockDoor: { name: 'Mock Door' }
      }}
      roles={{
        mockRole: { name: 'Mock Role' }
      }}
      employees={{
        mockEmployee: { name: 'Mock Employee' }
      }}
      requestsKeys={['mock']}
    />
  )

  expect(EnzymeToJson(component)).toMatchSnapshot()
})

it('renders without crashing with request and deleted data', () => {
  const mockDate = new Date('11/12/1994')

  const component = mount(
    <Log
      requests={{
        mock: {
          date: mockDate,
          employee: 'mockEmployee',
          role: 'mockRole',
          door: 'mockDoor',
          allowed: false,
        }
      }}
      doors={{}}
      roles={{}}
      employees={{}}
      requestsKeys={['mock']}
    />
  )

  expect(EnzymeToJson(component)).toMatchSnapshot()
})
