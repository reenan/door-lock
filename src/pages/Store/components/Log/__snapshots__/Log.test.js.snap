// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<Log
  doors={Object {}}
  employees={Object {}}
  requests={Object {}}
  requestsKeys={Array []}
  roles={Object {}}
>
  <div>
    <Header
      inverted={true}
      size="small"
    >
      <div
        className="ui small inverted header"
      >
        Here you can check all requests that were made to open your doors.
      </div>
    </Header>
    <Segment
      className="logs-container"
      tertiary={true}
    >
      <div
        className="ui tertiary segment logs-container"
      >
        <p>
          Looks like no request was made yet. Go and try one!
        </p>
      </div>
    </Segment>
  </div>
</Log>
`;

exports[`renders without crashing with request and deleted data 1`] = `
<Log
  doors={Object {}}
  employees={Object {}}
  requests={
    Object {
      "mock": Object {
        "allowed": false,
        "date": 1994-11-12T02:00:00.000Z,
        "door": "mockDoor",
        "employee": "mockEmployee",
        "role": "mockRole",
      },
    }
  }
  requestsKeys={
    Array [
      "mock",
    ]
  }
  roles={Object {}}
>
  <div>
    <Header
      inverted={true}
      size="small"
    >
      <div
        className="ui small inverted header"
      >
        Here you can check all requests that were made to open your doors.
      </div>
    </Header>
    <Segment
      className="logs-container"
      tertiary={true}
    >
      <div
        className="ui tertiary segment logs-container"
      >
        <div
          key="mock"
        >
          <p>
            On 
            <span>
              12
              /
              10
              /
              1994
               
              02
              :
              00
            </span>
            , 
            <span
              className="deleted"
              key="0"
            >
              Deleted Employee
            </span>
            <span
              className="deleted"
              key="1"
            >
               (Deleted Role)
            </span>
             requested to open 
            <span
              className="deleted"
            >
              Deleted Door
            </span>
             and his/hers request was 
            <span
              className="red"
            >
              not accepted
            </span>
          </p>
        </div>
      </div>
    </Segment>
  </div>
</Log>
`;

exports[`renders without crashing with request and saved data 1`] = `
<Log
  doors={
    Object {
      "mockDoor": Object {
        "name": "Mock Door",
      },
    }
  }
  employees={
    Object {
      "mockEmployee": Object {
        "name": "Mock Employee",
      },
    }
  }
  requests={
    Object {
      "mock": Object {
        "allowed": true,
        "date": 1994-11-12T02:00:00.000Z,
        "door": "mockDoor",
        "employee": "mockEmployee",
        "role": "mockRole",
      },
    }
  }
  requestsKeys={
    Array [
      "mock",
    ]
  }
  roles={
    Object {
      "mockRole": Object {
        "name": "Mock Role",
      },
    }
  }
>
  <div>
    <Header
      inverted={true}
      size="small"
    >
      <div
        className="ui small inverted header"
      >
        Here you can check all requests that were made to open your doors.
      </div>
    </Header>
    <Segment
      className="logs-container"
      tertiary={true}
    >
      <div
        className="ui tertiary segment logs-container"
      >
        <div
          key="mock"
        >
          <p>
            On 
            <span>
              12
              /
              10
              /
              1994
               
              02
              :
              00
            </span>
            , 
            <span
              key="0"
            >
              Mock Employee
            </span>
            <span
              key="1"
            >
               (
              Mock Role
              )
            </span>
             requested to open 
            <span>
              Mock Door
            </span>
             and his/hers request was 
            <span
              className="green"
            >
              accepted
            </span>
          </p>
        </div>
      </div>
    </Segment>
  </div>
</Log>
`;
